torch_geometric/nn/conv/point_transformer_conv.py

```diff
@@ -0,0 +1,110 @@
+from typing import Union
+from torch_geometric.typing import OptTensor, PairOptTensor, PairTensor, Adj
+
+from torch import Tensor
+from torch_sparse import SparseTensor, set_diag
+from torch_geometric.nn.conv import MessagePassing
+from torch_geometric.utils import remove_self_loops, add_self_loops
+
+from torch.nn import Sequential, ReLU, Linear
+from torch.nn.functional import softmax
+
+from ..inits import reset
+
+
+class PointTransformerConv(MessagePassing):
+    r"""The PointTransformer layer from the `"Point Transformer"
+    <https://arxiv.org/abs/2012.09164>` paper
+
+    .. math::
+        \mathbf{x}^{\prime}_i =  \sum_{j \in
+        \mathcal{N}(i) \cup \{ i \}} \rho \left( \gamma \left( \phi
+        \left( \mathbf{x}_i \right) - \psi \left( \mathbf{x}_j \right) \right)
+        + \delta  \right) \odot \left( \alpha \left(  \mathbf{x}_j \right)
+        + \delta \right) ,
+
+    where :math:`\gamma` and :math:`\delta = \theta \left( \mathbf{p}_j -
+    \mathbf{p}_i \right)` denote neural networks, *.i.e.* MLPs, that
+    respectively map transformed node features and relative spatial
+    coordinates. :math:`\mathbf{P} \in \mathbb{R}^{N \times D}`
+    defines the position of each point.
+
+    Args:
+        in_channels (int): Size of each input sample features
+        out_channels (int) : size of each output sample features
+        add_self_loops (bool, optional): If set to :obj:`False`, will not add
+            self-loops to the input graph. (default: :obj:`True`)
+
+        pos_mlp_channels : (int, optional) size of intermediary channels in
+            points relative spatial coordinates computation.
+            (default: :obj:`64`)
+        attn_mlp_channels : (int, optional) factor size of intermediary
+            channels in transformed features mlp. (default: :obj:`2`)
+
+        **kwargs (optional): Additional arguments of
+            :class:`torch_geometric.nn.conv.MessagePassing`.
+    """
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 add_self_loops=True,
+                 pos_mlp_channels=64,
+                 attn_mlp_channels=2):
+        super(PointTransformerConv,
+              self).__init__(aggr='add')  # "Add" aggregation
+
+        self.delta = Sequential(Linear(3, pos_mlp_channels), ReLU(),
+                                Linear(pos_mlp_channels, out_channels))
+        self.gamma = Sequential(
+            Linear(out_channels, out_channels * attn_mlp_channels), ReLU(),
+            Linear(out_channels * attn_mlp_channels, out_channels))
+        self.alpha = Linear(in_channels, out_channels)
+        self.phi = Linear(in_channels, out_channels)
+        self.psi = Linear(in_channels, out_channels)
```

Comment:
Do you think we could have an API similar to [PointNetCov](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.PointNetConv). So `gamma` and `delta` would become `gloabl_nn` and `local_nn` respectively, and they can be passed to `__init__` as `torch.nn.Module` . Their default values could probably be the current networks of `gamma` and `delta`.

Reply:
I also thought about it, hesitating between putting the 5 in an API form or none at all, but doing it for gamma and delta only would be a good compromise indeed !

Reply:
@wsad1 I made it in the form of an API, however I am disatified with the current way I implemented it as "default". Indeed I need the out_channels parameters to define the nn so I can't define it in the arguments, but putting as default `None` and saying it is initialized just after bothers me. What would be the best practice here ?